Fix regular expression matching date patterns in date/datetime `<select/>`-based helpers, allowing support for other locales

mimmi20 opened the pull request:
This will escape some meta characters.

Fixes #160

Ocramius commented on test/View/Helper/FormDateSelectTest.php:
```diff
@@ -115,4 +115,32 @@ public function testGetElements(): void
+            $this->assertInstanceOf(Select::class, $subElement);
+        }
+
+        $this->assertCount(31, $elements[0]->getValueOptions());
```
I feel like this test isn't really readable: what are we testing here, specifically? 🤔

mimmi20 replied:
fixed

Ocramius reviewed (changes requested):
Few things to adjust:

 1. commit message does not explain what's going on - some more clarity needed there
 2. test case does not make the issue clear: if this test failed, I wouldn't understand why, nor what it represents

Ocramius commented on test/View/Helper/FormDateSelectTest.php:
```diff
@@ -115,4 +117,28 @@ public function testGetElements(): void
+        $this->helper->setLocale($locale);
+        $this->helper->setDateType($dateType);
+
+        $method = new ReflectionMethod($this->helper, 'parsePattern');
```
Hmm, this looks even worse than before - we shouldn't test internal methods via reflection.

Also here, when reading this test out of context, it is unclear what you are verifying.

Here's my understanding:

 * when using `es_CL`
 * and setting a short date format
 * when I parse the pattern
 * then the pattern has day, month, year

So far, so good, but what's the effect on the public-facing behavior of this element helper?

Ocramius replied:
As for the commit message: `laminas#160: escape regex-meta-character in AbstractFormDateSelect `still doesn't say much to the reader :|

mimmi20 replied:
I'm escaping two regex meta characters which are currently unescaped, like descriped in issue #160. I dont know how to test this without reflection as the functions which are causing the error are protected.

Xerkus replied:
This should be tested via public behavior. If there is no public behavior that exposes the bug then it is fine to fix without providing a test.

Ocramius replied:
> then it is fine to fix without providing a test.

No, it is not: it means we have dead/unreachable code then, and it should just be dropped or refactored instead.
Bugfixing without a valid test is just more maintenance burden for upstream maintainers.

Ocramius replied:
Checking this out (locally) to re-design the test, and try to trip the issue with `es_CL`.

Ocramius replied:
I've pushed faefb5bbe578a071b95f3a34a145940a28b2acee, which solves this discussion, and provides some context of what is going on (for posterity).

Meanwhile, thanks for the patience, @mimmi20!

mimmi20 committed:
escape regex-meta-character in AbstractFormDateSelect and FormDateTimeSelect

Signed-off-by: Thomas Müller <mimmi20@live.de>

Ocramius committed:
Tests for #160 and #184: showing broken datetime picker rendering in `es_CL`+short formats

This change introduces verbose XML-alike-structure verification, which checks:

 1. that the HTML generated by `FormDateSelect` and `FormDateTimeSelect`
    matches the requested locale (we had zero tests verifying the generated HTML structure!)
 2. that the HTML generation actually works (fails on `3.4.x` / 0ca41651cee8486cef68b6e0281ef9be99379c78 )

The underlying issue was that the date format was being parsed by a bogus regular expression (fixed
in previous commit), which in turn led to a PHP warning being raised, as well as HTML rendering being
completely broken.

This test validates the work done by @mimmi20 in #184, and also highlights how antiquate
these HTML select elements are (please use `<input type="datetime-local"/>` or similar instead).

Signed-off-by: Marco Pivetta <ocramius@gmail.com>
